### PR TITLE
Pointer

### PR DESCRIPTION
**Pull request by latot**
Hi well, continue the conversation of https://github.com/hrydgard/ppsspp/pull/8550, trying to avoid non-portable code an option was use the FloatBits union to work(idea of @unknownbrackets ), the intention is skip the most posible transform pointers of differents types like (rd is float):

```
u32 *rdu = (u32 *)rd;
```

@unknownbrackets say functions like ReadVector should ideally only read float values.

```
We're explicitly wanting to look at the bits stored in that float and work with them individually. It might 
make sense to use a union instead, and in some places we do.... but I don't think ReadVector should 
necessarily read into something other than a float.

-[Unknown]
```

but actually this functions use float and u32 values transforming between it. .

Basically this do, before, some float enter to ReadVector, then ReadVector use float, and in some part convert it to u32 and use it.
now, some FloatBits enter to ReadVector, from there acces to the float and u32 value.

FloatBits was already implementd in code, (only in some parts), now its only a generalization of it, and i move it to the header.

### CODE REVIEW

**Comment by latot** [comment]
maybe remove 
void WriteVector(const float *rs, VectorSize N, int reg);
void ReadVector(float *rd, VectorSize N, int reg);
???

**Comment by hrydgard** [comment]
I'm not sure I like this change. It doesn't really make the code easier to read, and it doesn't provide a performance benefit, and what we have works on all our platforms...

**Comment by latot** [comment]
you are right, this its only to better portability, personally i think help to use more easy the functions, don't need change types or select other things, only pass the var.

**Comment by latot** [comment]
thinking, what do you think include FloatBits to the commontypes?(adding doubles)
checking are some similar issues with similar solution....

**Comment by unknownbrackets** [comment]
> ```diff
> -			WriteVector((const float*)Memory::GetPointer(addr), V_Quad, vt);
> +			FloatBits mgpr;
> +			mgpr._u8 = *Memory::GetPointer(addr);
> +			ReadVector(&mgpr, V_Quad, vt);
> ```

This is definitely wrong.

> but actually this functions use float and u32 values transforming between it. .

I'm not sure what you mean by this, but just to say again: in all of the cases where ReadVector/WriteVector are called, we're reinterpreting bits (like `reinterpret_cast`), not transforming anything.

This is really a very specific use case: we're emulating a processor, which will reinterpret the bits in the registers depending on what instructions you use.  HLE code is not going to have much reason to ever do this.  In general, GE code won't either (there are a few rare exceptions.)

I don't think we have any hope of targeting a platform where float is not 4 bits, and this will break horribly on such a platform either way.  This code presumably was already working on big endian as well, at least to some degree in the Xbox 360 port.

Is there a specific portability concern you're talking about?

-[Unknown]

**Comment by latot** [comment]
mm, its basically this line i wan't to avoid:
```
u32 *rdu = (u32 *)rd;
```
exist in ReadVector and WriteVector.

**Comment by hrydgard** [comment]
The thing we do in ReadVector and WriteVector was just an attempt to use integer instructions for copying instead of fp instructions. I think the impact is pretty much zero, and I'm getting rid of it shortly, which will remove the need for this.

**Comment by latot** [comment]
Oidokis, Thx!
